### ISSUE TITLE
Custom schedules for a fused op?

### ISSUE THREAD

**Issue by masahi** [dmlc/nnvm#275]
Hi, is there a way to inject a custom schedule into a fused op, such as Conv + BatchNorm + Relu?

NNVM generates the following fused TVM IR for my Conv + BatchNorm + Relu ops after tvm.lower(...). 
I'm working on a CPU schedule here, so there are things like parallel, and vectorize.

I'm not completely happy about the code below for the following reasons.
* BatchNorm + Relu are not fused into the Conv inner loop.
* BatchNorm + Relu compute is not parallelized. 

Ideally, I want to move BatchNorm + Relu into the Conv inner loop by injecting compute_at(), or at least parallelize the outer loop of BatchNorm + Relu. I also want to vectorize BatchNorm + Relu op, if possible. Let me know if there is a way to inject a custom schedule after fusion is done. 

Thanks
 
```
func_name: fuse_conv2d_broadcast_mul_broadcast_add_relu_2
// attr [pad_temp] storage_scope = "global"
allocate pad_temp[float32 * 12 * 128 * 66 * 66]
// attr [compute] storage_scope = "global"
allocate compute[float32 * 12 * 64 * 64 * 64]
produce pad_temp {
  parallel (i0.i1.fused, 0, 1536) {
    for (i2, 0, 66) {
      for (i3, 0, 66) {
        pad_temp[((((i0.i1.fused*66) + i2)*66) + i3)] = tvm_if_then_else(((((1 <= i2) && (i2 < 65)) && (1 <= i3)) && (i3 < 65)), input0[(((((i0.i1.fused*64) + i2)*64) + i3) + -65)], 0.000000f)
      }
    }
  }
}
produce compute {
  parallel (nn.ff.fused, 0, 768) {
    for (yy, 0, 64) {
      for (xx.outer, 0, 4) {
        compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = x16(0.000000f)
        for (rc, 0, 128) {
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp(((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)), 1, 16)]*x16(input1[((((nn.ff.fused % 64)*128) + rc)*9)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 1), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 1)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 2), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 2)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 66), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 3)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 67), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 4)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 68), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 5)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 132), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 6)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 133), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 7)])))
          compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] = (compute[ramp((((((nn.ff.fused*64) + yy)*4) + xx.outer)*16), 1, 16)] + (pad_temp[ramp((((((((nn.ff.fused/64)*8448) + yy)*66) + (xx.outer*16)) + (rc*4356)) + 134), 1, 16)]*x16(input1[(((((nn.ff.fused % 64)*128) + rc)*9) + 8)])))
        }
      }
    }
  }
}
produce compute {
  for (i0, 0, 12) {
    for (i1, 0, 64) {
      for (i2, 0, 64) {
        for (i3, 0, 64) {
          compute[((((((i0*64) + i1)*64) + i2)*64) + i3)] = max((((compute[((((((i0*64) + i1)*64) + i2)*64) + i3)] + input2[i1])*input3[i1]) + input4[i1]), 0.000000f)
        }
      }
    }
  }
}
```

**Comment by masahi** [comment]
ok, I found a way to parallelize & vectorize the batch norm + Relu, but still don't know how to move batch norm + relu into the conv inner loop. I can't use compute_at() here, because the output of the fused op must be root.

**Comment by masahi** [comment]
I was able to fuse conv and batchnorm + relu by moving conv into batch norm + relu loop.
